src/Raven.Server/Documents/Handlers/Processors/Indexes/IndexHandlerProcessorForConvertAutoIndex.cs

```diff
@@ -0,0 +1,716 @@
+
+        var context = new AutoIndexConversionContext();
+
+        var sb = new StringBuilder();
+
+        csharpClassName = GenerateClassName(autoIndex);
+        var className = autoIndex.Collection is Constants.Documents.Collections.AllDocumentsCollection or Constants.Documents.Collections.EmptyCollection
+            ? "object"
+            : Inflector.Singularize(autoIndex.Collection);
+
+        sb
+            .Append($"public class {csharpClassName} : {typeof(AbstractIndexCreationTask).FullName}<{className}");
+
+        if (autoIndex.Type == IndexType.AutoMapReduce)
+            sb.Append($", {csharpClassName}.Result");
+
+        sb
+            .AppendLine(">")
+            .AppendLine("{")
+            .AppendLine($"public {csharpClassName}()")
+            .AppendLine("{");
+
+        ConstructMap(autoIndex, sb, context);
+        ConstructReduce(autoIndex, sb);
+        ConstructFieldOptions(autoIndex, sb, context);
+
+        sb
+            .AppendLine("}");
+
+        if (autoIndex.Type == IndexType.AutoMapReduce)
+        {
+            sb
+                .AppendLine()
+                .AppendLine("public class Result")
+                .AppendLine("{");
+
+            foreach (var kvp in autoIndex.MapFields)
+            {
+                var fieldNames = GenerateFieldName(kvp.Key, kvp.Value.Indexing);
+
+                foreach (var f in fieldNames)
+                {
+                    switch (kvp.Value.Aggregation)
+                    {
+                        case AggregationOperation.Sum:
+                        case AggregationOperation.Count:
+                            {
+                                sb.AppendLine($"public int {f.FieldName} {{ get; set; }}");
+                                break;
+                            }
+                        default:
+                            throw new ArgumentOutOfRangeException();
+                    }
+                }
+            }
+
+            foreach (var kvp in autoIndex.GroupByFields)
+            {
+                var fieldNames = GenerateFieldName(kvp.Key, kvp.Value.Indexing);
+
+                foreach (var f in fieldNames)
+                {
+                    if (f.IsConstant)
+                        continue;
+
+                    sb.AppendLine($"public object {f.FieldName} {{ get; set; }}");
+                }
+            }
+
+            sb.AppendLine("}");
+        }
+
+        sb
+            .AppendLine("}");
+
+        using (var workspace = new AdhocWorkspace())
+        {
+            var syntaxTree = SyntaxFactory
+                .ParseSyntaxTree(sb.ToString());
+
+            var result = Formatter.Format(syntaxTree.GetRoot(), workspace);
+            return result.ToString();
+        }
+
+
+        static void ConstructMap(AutoIndexDefinition autoIndex, StringBuilder sb, AutoIndexConversionContext context)
+        {
+            sb
+                .AppendLine("Map = items => from item in items")
```

Comment:
Don't we need to specify collection name e.g. `in docs.Items`? where `Items` is a given collection name

Reply:
no, this is LINQ syntax